### PR TITLE
feat: filter layers by date

### PR DESCRIPTION
**Pull request by blacha**
This adds `?date[before]=` and `?date[after]=` to a number of routes to filter the imagery down to a specific date rate.

These dates are inclusive so `?date[before]=2022` its a `less than or equal`.

### CODE REVIEW

File: packages/lambda-tiler/src/util/config.loader.ts [hunk 1]
```diff
@@ -14,9 +14,9 @@ export class ConfigLoader {
   static async getDefaultConfig(): Promise<BasemapsConfigProvider> {
     const config = getDefaultConfig();
     if (config.assets == null) {
-      const cb = await config.ConfigBundle.get(config.ConfigBundle.id('latest'));
-      if (cb == null) throw new LambdaHttpResponse(400, 'Unable to get lastest config bundle for asset.');
-      config.assets = cb.assets;
+      // const cb = await config.ConfigBundle.get(config.ConfigBundle.id('latest'));
```
Comment (Wentao-Kuang) [inline packages/lambda-tiler/src/util/config.loader.ts:17]: Why do we comment out this? config assets is broken?

File: packages/lambda-tiler/src/util/filter.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  const dateBeforeQuery = req.query.get(FilterNames.DateBefore);
+
+  if (dateAfterQuery == null && dateBeforeQuery == null) return layers;
+  const dateAfter = parseDateAsIso(dateAfterQuery);
+  const dateBefore = parseDateAsIso(dateBeforeQuery);
+
+  return layers.filter((l) => {
+    if (l.title == null) return false;
+    const yearRange = extractYearRangeFromTitle(l.title);
+    if (yearRange == null) return false;
```
Comment (Wentao-Kuang) [inline packages/lambda-tiler/src/util/filter.ts:44]: Does the date range slider will ignore all the layers without title or year range? Shall we just make sure all the layers to have a validate title with year in it when we bundle the config and validate it?

File: packages/lambda-tiler/src/wmts.capability.ts [hunk 3]
```diff
@@ -29,7 +29,7 @@ export interface WmtsCapabilitiesParams {
   /** List of tile matrixes to output */
   tileMatrix: TileMatrixSet[];
   /** Should WMTS Layers be created for each imagery set inside this tileSet */
-  isIndividualLayers: boolean;
+  // isIndividualLayers: boolean;
```
Comment (Wentao-Kuang) [inline packages/lambda-tiler/src/wmts.capability.ts:32]: Comment out code.

File: packages/lambda-tiler/src/wmts.capability.ts [hunk 4]
```diff
@@ -66,17 +68,19 @@ export class WmtsCapabilities {
...
   maxZoom = 32;
+  layers: ConfigLayer[] | null | undefined;
 
   constructor(params: WmtsCapabilitiesParams) {
     this.httpBase = params.httpBase;
     this.provider = params.provider;
     this.tileSet = params.tileSet;
     this.config = params.config;
-    this.isIndividualLayers = params.isIndividualLayers;
+    // this.isIndividualLayers = params.isIndividualLayers;
```
Comment (Wentao-Kuang) [inline packages/lambda-tiler/src/wmts.capability.ts:78]: comment out code.

File: packages/shared/src/util.ts [hunk 5]
```diff
@@ -46,6 +46,20 @@ export function extractYearRangeFromName(name: string): [number, number] {
...
+  const yearChunk = t.slice(t.lastIndexOf(' ') + 1).trim(); // "(2019)" or "(2019-2020)"
+  if (yearChunk == null) return null;
+  if (!yearChunk.endsWith(')') || !yearChunk.startsWith('(')) return null;
+
+  const chunks = yearChunk
+    .slice(1, yearChunk.length - 1)
+    .split('-')
+    .map(Number);
+  if (chunks.length === 1 && !isNaN(chunks[0])) return chunks as [number];
+  else if (chunks.length === 2 && !isNaN(chunks[0]) && !isNaN(chunks[0])) return chunks as [number, number];
```
Comment (Wentao-Kuang) [inline packages/shared/src/util.ts:59]: The second `!isNaN(chunks[0]` should be `!isNaN(chunks[1]`.

File: packages/lambda-tiler/src/util/config.loader.ts (hunk 1 above)
Comment (blacha) [reply packages/lambda-tiler/src/util/config.loader.ts:17]: running a server without this commented out, causes the server to crash unless you specify a assets and you dont need the assets for most actions of the server.

but this shouldnt be left commented out.

File: packages/lambda-tiler/src/util/filter.ts (hunk 2 above)
Comment (blacha) [reply packages/lambda-tiler/src/util/filter.ts:44]: this is only applied when the `date[before]` or `date[after]` is found

**Review by Wentao-Kuang — APPROVED** [review]

**Review by Wentao-Kuang — DISMISSED** [review]

**Review by Wentao-Kuang — APPROVED** [review]

**Pull request merged**